Documentation Doxygéne/man/man3/MissionControl.3
.TH "MissionControl" 3 "Wed Sep 11 2013" "MARCS" \" -*- nroff -*-
.ad l
.nh
.SH NAME
MissionControl \- 
.SH SYNOPSIS
.br
.PP
.PP
\fC#include 'MissionControl\&.h'\fP
.PP
Inherits QObject\&.
.SS "Public Slots"
.SS "Signals"
.SS "Static Public Member Functions"
.SS "Private Member Functions"
.SS "Private Attributes"
.SS "Static Private Attributes"
.SH "Detailed Description"
.PP 
.SH "Constructor & Destructor Documentation"
.PP 
.SS "\fBMissionControl\fP ()\fC [private]\fP"

.PP
Constructor\&. 
.PP
\fBParameters:\fP
.RS 4
\fInone\fP 
.RE
.PP
\fBExceptions:\fP
.RS 4
\fInone\fP 
.RE
.PP
\fBReturns:\fP
.RS 4
\fBMissionControl\fP instance 
.RE
.PP

.PP
.nf
6 {
7         //m_pPlan = new FlightPlan();
8         //m_pMap = new Map();
9         //m_pMap->setFlightPlan(m_pPlan);
10 
11     m_pLaunchTimer\&.setInterval(1000);
12 
13     connect(CommunicationControl::getInstance(), SIGNAL(XBeeStatus(bool)), this, SLOT(updateXBee(bool)));
14     connect(PeripheralController::getInstance(), SIGNAL(directInputProblem()), this, SIGNAL(directXProblem()));
15     connect(PeripheralController::getInstance(), SIGNAL(dataFormatProblem()), this, SIGNAL(directXProblem()));
16     connect(PeripheralController::getInstance(), SIGNAL(deviceEnumProblem()), this, SIGNAL(directXProblem()));
17     connect(PeripheralController::getInstance(), SIGNAL(joystickProblem()), this, SIGNAL(controllerProblem()));
18     connect(PeripheralController::getInstance(), SIGNAL(noJoystick()), this, SIGNAL(controllerProblem()));
19     connect(PeripheralController::getInstance(), SIGNAL(controllerConnected()), this, SIGNAL(controllerConnected()));
20 
21     connect(CommunicationControl::getInstance(), SIGNAL(geolocation(double, double, double, double)), 
22         RPA::getInstance(), SLOT(geolocation(double, double, double, double)));
23     connect(CommunicationControl::getInstance(), SIGNAL(updateHeading(double)), RPA::getInstance(), SLOT(updateHeading(double)));
24     connect(CommunicationControl::getInstance(), SIGNAL(updateHeight(double)), RPA::getInstance(), SLOT(updateHeight(double)));
25     connect(CommunicationControl::getInstance(), SIGNAL(updateGPSQuality(int)), this, SLOT(updateGPS(int)));
26     connect(CommunicationControl::getInstance(), SIGNAL(batteryLevel(double)), this, SLOT(updateBattery(double)));
27     connect(CommunicationControl::getInstance(), SIGNAL(updateStatus(short)), this, SLOT(updateRadiocommand(short)));
28     connect(CommunicationControl::getInstance(), SIGNAL(inFlight(char)), this, SLOT(inFlight(char)));
29     connect(CommunicationControl::getInstance(), SIGNAL(waypointNAVInfo(char, short, short)), 
30         this, SLOT(updateNAV(char, short, short)));
31     connect(CommunicationControl::getInstance(), SIGNAL(rcData(short, short, short, short, short, short, char)), 
32         this, SLOT(rcData(short, short, short, short, short, short, char)));
33 
34     connect(RPA::getInstance(), SIGNAL(positionChanged()), this, SLOT(updateDigression()));
35     connect(&m_pWaypointTimer, SIGNAL(timeout()), this, SLOT(sendWaypoint()));
36     connect(&m_pLaunchTimer, SIGNAL(timeout()), this, SIGNAL(sendLaunch()));
37 
38     connect(CommunicationControl::getInstance(), SIGNAL(displayAck(char*)), this, SLOT(acknowledge(char*)));
39     connect(this, SIGNAL(sendLaunch()), CommunicationControl::getInstance(), SIGNAL(sendLaunch()));
40     connect(this, SIGNAL(move(double, double, double, double)), CommunicationControl::getInstance(), 
41         SIGNAL(sendMove(double, double, double, double)));
42 
43     m_bFirstWaypoint = true;
44     m_bNextSent = false;
45     m_bFPLaunched = false;
46     m_bFirstRC = true;
47 }
.fi
.SS "~\fBMissionControl\fP ()\fC [private]\fP"

.PP
Destructor\&. 
.PP
\fBParameters:\fP
.RS 4
\fInone\fP 
.RE
.PP
\fBExceptions:\fP
.RS 4
\fInone\fP 
.RE
.PP
\fBReturns:\fP
.RS 4
none 
.RE
.PP

.PP
.nf
59 {
60 }
.fi
.SH "Member Function Documentation"
.PP 
.SS "void acknowledge (char *m_pAck)\fC [slot]\fP"

.PP
Receive MUAV acknowledge\&. 
.PP
\fBParameters:\fP
.RS 4
\fIm_pAck\fP acknowledge string char* 
.RE
.PP
\fBExceptions:\fP
.RS 4
\fInone\fP 
.RE
.PP
\fBReturns:\fP
.RS 4
none 
.RE
.PP

.PP
.nf
185 {
186     QString ack(m_pAck);
187 
188     if (ack == WAYPOINT_ACK)
189     {
190         m_pWaypointTimer\&.stop();
191     }
192     else if (ack == DEFINE_HOME_ACK)
193     {
194         m_pLaunchTimer\&.stop();
195     }
196 }
.fi
.SS "void addMark ()\fC [slot]\fP"

.PP
.nf
313 {
314         //Waypoint* pWaypoint = new Waypoint(RPA::getInstance()->getCoordinates(), RPA::getInstance()->getHeight());
315         //pWaypoint->setNumber(MissionControl::getInstance()->getPlan()->getWaypointList()\&.size() +
316         //  MissionControl::getInstance()->getPlan()->getMarkList()\&.size() + 1);
317         //pWaypoint->setAction(new Action(GOTO));
318 
319         //MissionControl::getInstance()->getPlan()->addMark(pWaypoint);
320 }
.fi
.SS "void batteryLevel (doublep_pValue)\fC [signal]\fP"

.PP
Battery level\&. 
.PP
\fBParameters:\fP
.RS 4
\fInone\fP 
.RE
.PP
\fBExceptions:\fP
.RS 4
\fInone\fP 
.RE
.PP
\fBReturns:\fP
.RS 4
none 
.RE
.PP

.SS "void cannotLaunchFP (boolp_bBattery, boolp_bGPS, boolp_bXBee, boolp_bStatus, boolp_bFlightPlan, boolp_bEngine)\fC [signal]\fP"

.PP
Cannot launch flight plan event\&. 
.PP
\fBParameters:\fP
.RS 4
\fIp_bBattery\fP battery level above low bool 
.br
\fIp_bGPS\fP enough satellites bool 
.br
\fIp_bXBee\fP link through XBee bool 
.br
\fIp_bStatus\fP MUAV in GPS mode bool 
.br
\fIp_bFlightPlan\fP flight plan loaded bool 
.br
\fIp_bEngine\fP engines started bool 
.RE
.PP
\fBExceptions:\fP
.RS 4
\fInone\fP 
.RE
.PP
\fBReturns:\fP
.RS 4
none 
.RE
.PP

.SS "void controllerConnected ()\fC [signal]\fP"

.PP
Controller connected\&. 
.PP
\fBParameters:\fP
.RS 4
\fInone\fP 
.RE
.PP
\fBExceptions:\fP
.RS 4
\fInone\fP 
.RE
.PP
\fBReturns:\fP
.RS 4
none 
.RE
.PP

.SS "void controllerProblem ()\fC [signal]\fP"

.PP
Problem with the game controller\&. 
.PP
\fBParameters:\fP
.RS 4
\fInone\fP 
.RE
.PP
\fBExceptions:\fP
.RS 4
\fInone\fP 
.RE
.PP
\fBReturns:\fP
.RS 4
none 
.RE
.PP

.SS "void digression (doublep_pValue)\fC [signal]\fP"

.PP
Digression value\&. 
.PP
\fBParameters:\fP
.RS 4
\fInone\fP 
.RE
.PP
\fBExceptions:\fP
.RS 4
\fInone\fP 
.RE
.PP
\fBReturns:\fP
.RS 4
none 
.RE
.PP

.SS "void directXProblem ()\fC [signal]\fP"

.PP
Direct X problem\&. 
.PP
\fBParameters:\fP
.RS 4
\fInone\fP 
.RE
.PP
\fBExceptions:\fP
.RS 4
\fInone\fP 
.RE
.PP
\fBReturns:\fP
.RS 4
none 
.RE
.PP

.SS "void down ()\fC [slot]\fP"

.PP
Decrease altitude command\&. 
.PP
\fBParameters:\fP
.RS 4
\fInone\fP 
.RE
.PP
\fBExceptions:\fP
.RS 4
\fInone\fP 
.RE
.PP
\fBReturns:\fP
.RS 4
none 
.RE
.PP

.PP
.nf
308 {
309     emit move(0, 0, RPA::getInstance()->getAltitude() - 10, RPA::getInstance()->getHeading());
310 }
.fi
.SS "void endFlightPlan ()\fC [signal]\fP"

.PP
End of flight plan event\&. 
.PP
\fBParameters:\fP
.RS 4
\fInone\fP 
.RE
.PP
\fBExceptions:\fP
.RS 4
\fInone\fP 
.RE
.PP
\fBReturns:\fP
.RS 4
none 
.RE
.PP

.SS "void flightPlanLaunched ()\fC [signal]\fP"

.PP
Flight plan launched acknowledge\&. 
.PP
\fBParameters:\fP
.RS 4
\fInone\fP 
.RE
.PP
\fBExceptions:\fP
.RS 4
\fInone\fP 
.RE
.PP
\fBReturns:\fP
.RS 4
none 
.RE
.PP

.SS "\fBMissionControl\fP * getInstance ()\fC [static]\fP"

.PP
\fBMissionControl\fP lone instance getter\&. 
.PP
\fBParameters:\fP
.RS 4
\fInone\fP 
.RE
.PP
\fBExceptions:\fP
.RS 4
\fInone\fP 
.RE
.PP
\fBReturns:\fP
.RS 4
lone instance MissionControl* 
.RE
.PP

.PP
.nf
63 {
64     if (singleton == NULL)
65     {
66         singleton = new MissionControl();
67     }
68 
69     return singleton;
70 }
.fi
.SS "void GPSLevel (intp_pValue)\fC [signal]\fP"

.PP
GPS reception level\&. 
.PP
\fBParameters:\fP
.RS 4
\fInone\fP 
.RE
.PP
\fBExceptions:\fP
.RS 4
\fInone\fP 
.RE
.PP
\fBReturns:\fP
.RS 4
none 
.RE
.PP

.SS "void inFlight (charp_cValue)\fC [slot]\fP"

.PP
Define if the MUAV has its engines started\&. 
.PP
\fBParameters:\fP
.RS 4
\fInone\fP 
.RE
.PP
\fBExceptions:\fP
.RS 4
\fInone\fP 
.RE
.PP
\fBReturns:\fP
.RS 4
none 
.RE
.PP

.PP
.nf
268 {
269     m_cEngine = p_cValue;
270 }
.fi
.SS "void kill ()\fC [static]\fP"

.PP
Instance killer\&. 
.PP
\fBParameters:\fP
.RS 4
\fInone\fP 
.RE
.PP
\fBExceptions:\fP
.RS 4
\fInone\fP 
.RE
.PP
\fBReturns:\fP
.RS 4
none 
.RE
.PP

.PP
.nf
50 {
51     if (singleton != NULL)
52     {
53         delete singleton;
54         singleton = NULL;
55     }
56 }
.fi
.SS "void launchFlightPlan ()\fC [slot]\fP"

.PP
Launch the flight plan\&. 
.PP
\fBParameters:\fP
.RS 4
\fInone\fP 
.RE
.PP
\fBExceptions:\fP
.RS 4
\fIindicate\fP if cannot launch (bad parameters) 
.RE
.PP
\fBReturns:\fP
.RS 4
none 
.RE
.PP

.PP
.nf
199 {
200         //bool bFlightPlanDefined = m_pPlan->getWaypointList()\&.size() != 0;
201 
202         //if (m_dBattery > batteryVeryLow && /*m_nGPS > GPSVeryLow &&*/ m_bXBee
203         //  && (m_nStatus & GPS) && bFlightPlanDefined && m_cEngine == 1)
204     if (true)
205     {
206         m_bFPLaunched = true;
207         emit flightPlanLaunched();
208     }
209     else
210     {
211                 //emit cannotLaunchFP(m_dBattery > batteryVeryLow, m_nGPS > GPSVeryLow, m_bXBee,
212                 //  m_nStatus & GPS, bFlightPlanDefined, m_cEngine == 1);
213     }
214 }
.fi
.SS "void move (doublex, doubley, doublez, doubleyaw)\fC [signal]\fP"

.PP
\fBWaypoint\fP to send\&. 
.PP
\fBParameters:\fP
.RS 4
\fIx\fP in meters double 
.br
\fIy\fP in meters double 
.br
\fIz\fP altitude double 
.br
\fIyaw\fP heading double 
.RE
.PP
\fBExceptions:\fP
.RS 4
\fInone\fP 
.RE
.PP
\fBReturns:\fP
.RS 4
none 
.RE
.PP

.SS "void moveBackward ()\fC [slot]\fP"

.PP
Move backward command\&. 
.PP
\fBParameters:\fP
.RS 4
\fInone\fP 
.RE
.PP
\fBExceptions:\fP
.RS 4
\fInone\fP 
.RE
.PP
\fBReturns:\fP
.RS 4
none 
.RE
.PP

.PP
.nf
278 {
279     emit move(-10, 0, RPA::getInstance()->getAltitude(), RPA::getInstance()->getHeading());
280 }
.fi
.SS "void moveForward ()\fC [slot]\fP"

.PP
Move forward command\&. 
.PP
\fBParameters:\fP
.RS 4
\fInone\fP 
.RE
.PP
\fBExceptions:\fP
.RS 4
\fInone\fP 
.RE
.PP
\fBReturns:\fP
.RS 4
none 
.RE
.PP

.PP
.nf
273 {
274     emit move(10, 0, RPA::getInstance()->getAltitude(), RPA::getInstance()->getHeading());
275 }
.fi
.SS "void moveLeft ()\fC [slot]\fP"

.PP
Move left command\&. 
.PP
\fBParameters:\fP
.RS 4
\fInone\fP 
.RE
.PP
\fBExceptions:\fP
.RS 4
\fInone\fP 
.RE
.PP
\fBReturns:\fP
.RS 4
none 
.RE
.PP

.PP
.nf
283 {
284     emit move(0, 10, RPA::getInstance()->getAltitude(), RPA::getInstance()->getHeading());
285 }
.fi
.SS "void moveRight ()\fC [slot]\fP"

.PP
Move right command\&. 
.PP
\fBParameters:\fP
.RS 4
\fInone\fP 
.RE
.PP
\fBExceptions:\fP
.RS 4
\fInone\fP 
.RE
.PP
\fBReturns:\fP
.RS 4
none 
.RE
.PP

.PP
.nf
288 {
289     emit move(0, -10, RPA::getInstance()->getAltitude(), RPA::getInstance()->getHeading());
290 }
.fi
.SS "bool nextWaypoint ()\fC [slot]\fP"

.PP
Next waypoint index increment\&. 
.PP
\fBParameters:\fP
.RS 4
\fInone\fP 
.RE
.PP
\fBExceptions:\fP
.RS 4
\fInone\fP 
.RE
.PP
\fBReturns:\fP
.RS 4
end of flight plan reached 
.RE
.PP

.PP
.nf
133 {
134     bool ret = false;
135         //int index = m_pPlan->getNextWaypointIndex();
136 
137         //if (index < m_pPlan->getWaypointList()\&.size())
138         if(true)
139     {
140         //  m_pPlan->setNextWaypointIndex(index + 1);
141         ret = true;
142     }
143     else
144     {
145         ret = false;
146     }
147 
148         //m_pMap->displayWaypoints();
149     emit resetWaypointNotification();
150 
151     return ret;
152 }
.fi
.SS "void pauseFlightPlan ()\fC [slot]\fP"

.PP
Pause flight plan (no reset of next waypoint index) 
.PP
\fBParameters:\fP
.RS 4
\fInone\fP 
.RE
.PP
\fBExceptions:\fP
.RS 4
\fInone\fP 
.RE
.PP
\fBReturns:\fP
.RS 4
none 
.RE
.PP

.PP
.nf
224 {
225     m_bFPLaunched = false;
226 }
.fi
.SS "void radiocommandConnection (boolp_pValue)\fC [signal]\fP"

.PP
Radiocommand connection state\&. 
.PP
\fBParameters:\fP
.RS 4
\fInone\fP 
.RE
.PP
\fBExceptions:\fP
.RS 4
\fInone\fP 
.RE
.PP
\fBReturns:\fP
.RS 4
none 
.RE
.PP

.SS "void rcData (shortp_nYaw, shortp_nPitch, shortp_nRoll, shortp_nThrust, shortp_nSerial, shortp_nGPSHeightControl, charp_cValid)\fC [slot]\fP"

.PP
Define if remote control values changed and update them\&. 
.PP
\fBParameters:\fP
.RS 4
\fIyaw\fP short 
.br
\fIpitch\fP short 
.br
\fIroll\fP short 
.br
\fIthrust\fP short 
.br
\fIserial\fP short 
.br
\fIGPSHeightControl\fP short 
.br
\fIvalid\fP char 
.RE
.PP
\fBExceptions:\fP
.RS 4
\fInone\fP 
.RE
.PP
\fBReturns:\fP
.RS 4
none 
.RE
.PP

.PP
.nf
241 {
242     if (p_cValid)
243     {
244         if (!m_bFirstRC && PeripheralController::getInstance()->isMouseMode())
245         {
246             if (m_pPreviousYawRC != p_nYaw || m_pPreviousPitchRC != p_nPitch
247                 || m_pPreviousRollRC != p_nRoll || m_pPreviousThrustRC != p_nThrust
248                 || m_pPreviousSerialRC != p_nSerial || m_pPreviousGPSHeightControlRC != p_nGPSHeightControl)
249             {
250                 emit RCDataChanged();
251 
252                 m_pPreviousYawRC = p_nYaw;
253                 m_pPreviousPitchRC = p_nPitch;
254                 m_pPreviousRollRC = p_nRoll;
255                 m_pPreviousThrustRC = p_nThrust;
256                 m_pPreviousSerialRC = p_nSerial;
257                 m_pPreviousGPSHeightControlRC = p_nGPSHeightControl;
258             }
259         }
260         else
261         {
262             m_bFirstRC = false;
263         }
264     }
265 }
.fi
.SS "void RCDataChanged ()\fC [signal]\fP"

.PP
Remote control values changed\&. 
.PP
\fBParameters:\fP
.RS 4
\fInone\fP 
.RE
.PP
\fBExceptions:\fP
.RS 4
\fInone\fP 
.RE
.PP
\fBReturns:\fP
.RS 4
none 
.RE
.PP

.SS "void resetWaypointNotification ()\fC [signal]\fP"

.PP
Change next waypoint\&. 
.PP
\fBParameters:\fP
.RS 4
\fInone\fP 
.RE
.PP
\fBExceptions:\fP
.RS 4
\fInone\fP 
.RE
.PP
\fBReturns:\fP
.RS 4
none 
.RE
.PP

.SS "void sendLaunch ()\fC [signal]\fP"

.PP
Send launch order\&. 
.PP
\fBParameters:\fP
.RS 4
\fInone\fP 
.RE
.PP
\fBExceptions:\fP
.RS 4
\fInone\fP 
.RE
.PP
\fBReturns:\fP
.RS 4
none 
.RE
.PP

.SS "void sendWaypoint ()\fC [slot]\fP"

.PP
Send waypoint order\&. 
.PP
\fBParameters:\fP
.RS 4
\fInone\fP 
.RE
.PP
\fBExceptions:\fP
.RS 4
\fInone\fP 
.RE
.PP
\fBReturns:\fP
.RS 4
none 
.RE
.PP

.PP
.nf
120 {
121         //if (m_pPlan->getWaypointList()\&.size() > 0)
122     {
123                 //Waypoint* pWaypoint = m_pPlan->getWaypointList()[m_pPlan->getNextWaypointIndex()];
124                 //LatLongCoord pCoordinates = pWaypoint->getCoordinates();
125 
126                 //emit waypoint(pCoordinates\&.getLongitude(), pCoordinates\&.getLatitude(), pWaypoint->getAltitude(), pWaypoint->getNumber());
127 
128     }
129 m_pWaypointTimer\&.start(1000);
130 }
.fi
.SS "void stopFlightPlan ()\fC [slot]\fP"

.PP
Stop flight plan (reset of next waypoint index) 
.PP
\fBParameters:\fP
.RS 4
\fInone\fP 
.RE
.PP
\fBExceptions:\fP
.RS 4
\fInone\fP 
.RE
.PP
\fBReturns:\fP
.RS 4
none 
.RE
.PP

.PP
.nf
217 {
218     m_bFPLaunched = false;
219         //m_pPlan->setNextWaypointIndex(0);
220     m_bFirstWaypoint = true;
221 }
.fi
.SS "void turnLeft ()\fC [slot]\fP"

.PP
Turn left command\&. 
.PP
\fBParameters:\fP
.RS 4
\fInone\fP 
.RE
.PP
\fBExceptions:\fP
.RS 4
\fInone\fP 
.RE
.PP
\fBReturns:\fP
.RS 4
none 
.RE
.PP

.PP
.nf
293 {
294     emit move(0, 0, RPA::getInstance()->getAltitude(), RPA::getInstance()->getHeading() - 10);
295 }
.fi
.SS "void turnRight ()\fC [slot]\fP"

.PP
Turn right command\&. 
.PP
\fBParameters:\fP
.RS 4
\fInone\fP 
.RE
.PP
\fBExceptions:\fP
.RS 4
\fInone\fP 
.RE
.PP
\fBReturns:\fP
.RS 4
none 
.RE
.PP

.PP
.nf
298 {
299     emit move(0, 0, RPA::getInstance()->getAltitude(), RPA::getInstance()->getHeading() + 10);
300 }
.fi
.SS "void up ()\fC [slot]\fP"

.PP
Increase altitude command\&. 
.PP
\fBParameters:\fP
.RS 4
\fInone\fP 
.RE
.PP
\fBExceptions:\fP
.RS 4
\fInone\fP 
.RE
.PP
\fBReturns:\fP
.RS 4
none 
.RE
.PP

.PP
.nf
303 {
304     emit move(0, 0, RPA::getInstance()->getAltitude() + 10, RPA::getInstance()->getHeading());
305 }
.fi
.SS "void updateBattery (doublep_pValue)\fC [slot]\fP"

.PP
Battery level update\&. 
.PP
\fBParameters:\fP
.RS 4
\fIp_pValue\fP battery level double 
.RE
.PP
\fBExceptions:\fP
.RS 4
\fInone\fP 
.RE
.PP
\fBReturns:\fP
.RS 4
none 
.RE
.PP

.PP
.nf
77 {
78     emit batteryLevel(p_pValue);
79     m_dBattery = p_pValue;
80 }
.fi
.SS "void updateDigression ()\fC [slot]\fP"

.PP
Compute the digression value\&. 
.PP
\fBParameters:\fP
.RS 4
\fInone\fP 
.RE
.PP
\fBExceptions:\fP
.RS 4
\fInone\fP 
.RE
.PP
\fBReturns:\fP
.RS 4
none 
.RE
.PP

.PP
.nf
89 {
90     LatLongCoord* p_pPosition = RPA::getInstance()->getCoordinates();
91 

MARCS/MissionControl.cpp
#include "MissionControl.h"

MissionControl* MissionControl::singleton = NULL;

MissionControl::MissionControl()
{
        //m_pPlan = new FlightPlan();
        //m_pMap = new Map();
        //m_pMap->setFlightPlan(m_pPlan);

	m_pLaunchTimer.setInterval(1000);

	connect(CommunicationControl::getInstance(), SIGNAL(XBeeStatus(bool)), this, SLOT(updateXBee(bool)));
	connect(PeripheralController::getInstance(), SIGNAL(directInputProblem()), this, SIGNAL(directXProblem()));
	connect(PeripheralController::getInstance(), SIGNAL(dataFormatProblem()), this, SIGNAL(directXProblem()));
	connect(PeripheralController::getInstance(), SIGNAL(deviceEnumProblem()), this, SIGNAL(directXProblem()));
	connect(PeripheralController::getInstance(), SIGNAL(joystickProblem()), this, SIGNAL(controllerProblem()));
	connect(PeripheralController::getInstance(), SIGNAL(noJoystick()), this, SIGNAL(controllerProblem()));
	connect(PeripheralController::getInstance(), SIGNAL(controllerConnected()), this, SIGNAL(controllerConnected()));

	connect(CommunicationControl::getInstance(), SIGNAL(geolocation(double, double, double, double)), 
        RPA::getInstance(), SLOT(geolocation(double, double, double, double)));
    connect(CommunicationControl::getInstance(), SIGNAL(updateHeading(double)), RPA::getInstance(), SLOT(updateHeading(double)));
    connect(CommunicationControl::getInstance(), SIGNAL(updateHeight(double)), RPA::getInstance(), SLOT(updateHeight(double)));
	connect(CommunicationControl::getInstance(), SIGNAL(updateGPSQuality(int)), this, SLOT(updateGPS(int)));
	connect(CommunicationControl::getInstance(), SIGNAL(batteryLevel(double)), this, SLOT(updateBattery(double)));
	connect(CommunicationControl::getInstance(), SIGNAL(updateStatus(short)), this, SLOT(updateRadiocommand(short)));
	connect(CommunicationControl::getInstance(), SIGNAL(inFlight(char)), this, SLOT(inFlight(char)));
	connect(CommunicationControl::getInstance(), SIGNAL(waypointNAVInfo(char, short, short)), 
		this, SLOT(updateNAV(char, short, short)));
	connect(CommunicationControl::getInstance(), SIGNAL(rcData(short, short, short, short, short, short, char)), 
		this, SLOT(rcData(short, short, short, short, short, short, char)));

    connect(RPA::getInstance(), SIGNAL(positionChanged()), this, SLOT(updateDigression()));
	connect(&m_pWaypointTimer, SIGNAL(timeout()), this, SLOT(sendWaypoint()));
	connect(&m_pLaunchTimer, SIGNAL(timeout()), this, SIGNAL(sendLaunch()));

	connect(CommunicationControl::getInstance(), SIGNAL(displayAck(char*)), this, SLOT(acknowledge(char*)));
	connect(this, SIGNAL(sendLaunch()), CommunicationControl::getInstance(), SIGNAL(sendLaunch()));
	connect(this, SIGNAL(move(double, double, double, double)), CommunicationControl::getInstance(), 
		SIGNAL(sendMove(double, double, double, double)));

	m_bFirstWaypoint = true;
	m_bNextSent = false;
	m_bFPLaunched = false;
	m_bFirstRC = true;
}

void MissionControl::kill()
{
	if (singleton != NULL)
	{
		delete singleton;
		singleton = NULL;
	}
}

MissionControl::~MissionControl()
{
}

MissionControl* MissionControl::getInstance()
{
	if (singleton == NULL)
	{
		singleton = new MissionControl();
	}

	return singleton;
}





void MissionControl::updateBattery(double p_pValue)
{
	emit batteryLevel(p_pValue);
	m_dBattery = p_pValue;
}

void MissionControl::updateGPS(int p_pValue)
{
	emit GPSLevel(p_pValue);
	m_nGPS = p_pValue;
}

void MissionControl::updateDigression()
{
    LatLongCoord* p_pPosition = RPA::getInstance()->getCoordinates();

        //if (m_pPlan->getNextWaypointIndex() < m_pPlan->getWaypointList().size() && m_pPlan->getNextWaypointIndex() > 0)
	{
                //Waypoint* pFirstWaypoint = m_pPlan->getWaypointList()[m_pPlan->getNextWaypointIndex() - 1];
                //Waypoint* pSecondWaypoint = m_pPlan->getWaypointList()[m_pPlan->getNextWaypointIndex()];

                //if (pFirstWaypoint != NULL && pSecondWaypoint != NULL)
                    if (true)
		{
                        //LatLongCoord* p_pFirstWaypointPosition = pFirstWaypoint->getCoordinates();
                        //LatLongCoord* p_pSecondWaypointPosition = pSecondWaypoint->getCoordinates();
                        //emit digression(m_pMap->distanceFromLeg(p_pPosition, p_pFirstWaypointPosition, p_pSecondWaypointPosition));
		}
	}
}

void MissionControl::updateXBee(bool p_pValue)
{
	emit XBeeConnection(p_pValue);
	m_bXBee = p_pValue;
}

void MissionControl::updateRadiocommand(short p_pValue)
{
	emit radiocommandConnection(!(p_pValue & EMERGENCY));
	m_nStatus = p_pValue;
}

void MissionControl::sendWaypoint()
{
        //if (m_pPlan->getWaypointList().size() > 0)
	{
                //Waypoint* pWaypoint = m_pPlan->getWaypointList()[m_pPlan->getNextWaypointIndex()];
                //LatLongCoord pCoordinates = pWaypoint->getCoordinates();

                //emit waypoint(pCoordinates.getLongitude(), pCoordinates.getLatitude(), pWaypoint->getAltitude(), pWaypoint->getNumber());

    }
m_pWaypointTimer.start(1000);
}

bool MissionControl::nextWaypoint()
{
	bool ret = false;
        //int index = m_pPlan->getNextWaypointIndex();

        //if (index < m_pPlan->getWaypointList().size())
        if(true)
	{
        //	m_pPlan->setNextWaypointIndex(index + 1);
		ret = true;
	}
	else
	{
		ret = false;
	}

        //m_pMap->displayWaypoints();
	emit resetWaypointNotification();

	return ret;
}

void MissionControl::updateNAV(char p_cWaypointNumber, short p_nDistance, short p_cStatus)
{
	if (m_bFPLaunched)
	{
		if (m_bFirstWaypoint == true)
		{
			sendLaunch();
			sendWaypoint();
			m_bFirstWaypoint = false;
			m_pLaunchTimer.start();
		}
		else if (p_cStatus == WP_NAVSTAT_REACHED_POS && m_bNextSent == false)
		{
			if (nextWaypoint())
			{
				sendWaypoint();
				m_bNextSent = true;
			}
			else
			{
				emit endFlightPlan();
			}
		}
		else
		{
			m_bNextSent = false;
		}
	}
}

void MissionControl::acknowledge(char* m_pAck)
{
	QString ack(m_pAck);

	if (ack == WAYPOINT_ACK)
	{
		m_pWaypointTimer.stop();
	}
	else if (ack == DEFINE_HOME_ACK)
	{
		m_pLaunchTimer.stop();
	}
}

void MissionControl::launchFlightPlan()
{
        //bool bFlightPlanDefined = m_pPlan->getWaypointList().size() != 0;

        //if (m_dBattery > batteryVeryLow && /*m_nGPS > GPSVeryLow &&*/ m_bXBee
        //	&& (m_nStatus & GPS) && bFlightPlanDefined && m_cEngine == 1)
    if (true)
    {
		m_bFPLaunched = true;
		emit flightPlanLaunched();
	}
	else
	{
                //emit cannotLaunchFP(m_dBattery > batteryVeryLow, m_nGPS > GPSVeryLow, m_bXBee,
                //	m_nStatus & GPS, bFlightPlanDefined, m_cEngine == 1);
	}
}

void MissionControl::stopFlightPlan()
{
	m_bFPLaunched = false;
        //m_pPlan->setNextWaypointIndex(0);
	m_bFirstWaypoint = true;
}

void MissionControl::pauseFlightPlan()
{
	m_bFPLaunched = false;
}

void MissionControl::updateWayStatus(char p_cWaypointNumber, short p_nDistance, short p_nNavStatus)
{
	if (p_nNavStatus == (WP_NAVSTAT_REACHED_POS_TIME | WP_NAVSTAT_REACHED_POS | WP_NAVSTAT_20M))
	{
        //	if (m_pPlan->getNextWaypointIndex() < m_pPlan->getWaypointList().size() - 1)
		{
			emit waypointReached();
		}
	}
}

void MissionControl::rcData(short p_nYaw, short p_nPitch, short p_nRoll, short p_nThrust, 
	short p_nSerial, short p_nGPSHeightControl, char p_cValid)
{
	if (p_cValid)
	{
		if (!m_bFirstRC && PeripheralController::getInstance()->isMouseMode())
		{
			if (m_pPreviousYawRC != p_nYaw || m_pPreviousPitchRC != p_nPitch
				|| m_pPreviousRollRC != p_nRoll || m_pPreviousThrustRC != p_nThrust
				|| m_pPreviousSerialRC != p_nSerial || m_pPreviousGPSHeightControlRC != p_nGPSHeightControl)
			{
				emit RCDataChanged();

				m_pPreviousYawRC = p_nYaw;
				m_pPreviousPitchRC = p_nPitch;
				m_pPreviousRollRC = p_nRoll;
				m_pPreviousThrustRC = p_nThrust;
				m_pPreviousSerialRC = p_nSerial;
				m_pPreviousGPSHeightControlRC = p_nGPSHeightControl;
			}
		}
		else
		{
			m_bFirstRC = false;
		}
	}
}

void MissionControl::inFlight(char p_cValue)
{
	m_cEngine = p_cValue;
}

void MissionControl::moveForward()
{
    emit move(10, 0, RPA::getInstance()->getAltitude(), RPA::getInstance()->getHeading());
}

void MissionControl::moveBackward()
{
    emit move(-10, 0, RPA::getInstance()->getAltitude(), RPA::getInstance()->getHeading());
}

void MissionControl::moveLeft()
{
    emit move(0, 10, RPA::getInstance()->getAltitude(), RPA::getInstance()->getHeading());
}

void MissionControl::moveRight()
{
    emit move(0, -10, RPA::getInstance()->getAltitude(), RPA::getInstance()->getHeading());
}

void MissionControl::turnLeft()
{
    emit move(0, 0, RPA::getInstance()->getAltitude(), RPA::getInstance()->getHeading() - 10);
}

void MissionControl::turnRight()
{
    emit move(0, 0, RPA::getInstance()->getAltitude(), RPA::getInstance()->getHeading() + 10);
}

void MissionControl::up()
{
    emit move(0, 0, RPA::getInstance()->getAltitude() + 10, RPA::getInstance()->getHeading());
}

void MissionControl::down()
{
    emit move(0, 0, RPA::getInstance()->getAltitude() - 10, RPA::getInstance()->getHeading());
}

void MissionControl::addMark()
{
        //Waypoint* pWaypoint = new Waypoint(RPA::getInstance()->getCoordinates(), RPA::getInstance()->getHeight());
        //pWaypoint->setNumber(MissionControl::getInstance()->getPlan()->getWaypointList().size() +
        //	MissionControl::getInstance()->getPlan()->getMarkList().size() + 1);
        //pWaypoint->setAction(new Action(GOTO));

        //MissionControl::getInstance()->getPlan()->addMark(pWaypoint);
}
